src/crypto/ecdh.rs

```diff
@@ -145,18 +145,41 @@ pub fn decrypt(priv_key: &ECDHSecretKey, mpis: &[Mpi], fingerprint: &[u8]) -> Re
         .copy_from_slice(encrypted_session_key);
 
     let mut decrypted_key_padded = aes_kw::unwrap(&z, &encrypted_session_key_vec)?;
-    // PKCS5 unpadding (PKCS5 is PKCS7 with a blocksize of 8)
+    // PKCS5-style unpadding (PKCS5 is PKCS7 with a blocksize of 8).
+    //
+    // RFC 6637 describes the padding:
+    // a) "The result is padded using the method described in [PKCS5] to the 8-byte granularity."
+    // b) "For example, assuming that an AES algorithm is used for the session key, the sender MAY
+    // use 21, 13, and 5 bytes of padding for AES-128, AES-192, and AES-256, respectively, to
+    // provide the same number of octets, 40 total, as an input to the key wrapping method."
+    //
+    // So while the padding ensures that the length of the padded message is a multiple of 8, the
+    // padding may exceed 8 bytes in size.
     {
         let len = decrypted_key_padded.len();
         let block_size = 8;
         ensure!(len % block_size == 0, "invalid key length {}", len);
         ensure!(!decrypted_key_padded.is_empty(), "empty key is not valid");
 
-        // grab the last block
-        let offset = len - block_size;
-        let last_block = GenericArray::<u8, U8>::from_slice(&decrypted_key_padded[offset..]);
-        let unpadded_last_block = Pkcs7::unpad(last_block)?;
-        let unpadded_len = offset + unpadded_last_block.len();
+        // The last byte should contain the padding symbol, which is also the padding length
+        let pad = decrypted_key_padded.last().expect("is not empty");
```

Comment:
I'm just wondering if it's OK to panic in a library even if the data is malformed? :thinking: (I was thinking about using `return Err(Error::UnpadError);`). Just a random note. :wave:

Reply:
This code should probably be rewritten to be more readable. But that said, as far as I can tell, this `expect()` cannot trigger:

The `ensure!(!decrypted_key_padded.is_empty(), "empty key is not valid");` right above makes sure that there is at least one byte of data in `decrypted_key_padded`.

Reply:
Ok. That clarifies it for me, although it looked weird at first... :+1: (closing this thread)